src/ripple/consensus/Validations.h

```diff
@@ -518,211 +625,280 @@ class Validations
-
-                SeqType const seq = v.seq();
-                if ((seq == SeqType{0}) || (seq >= cutoffBefore))
-                {
-                    // contains a live record
-                    bool countPreferred =
-                        valCurrentLedger && (v.ledgerID() == currentLedger);
-
-                    if (!countPreferred &&  // allow up to one ledger slip in
-                                            // either direction
-                        ((valCurrentLedger &&
-                          (prevLedgerID == currentLedger)) ||
-                         (valPriorLedger && (v.ledgerID() == priorLedger))))
-                    {
-                        countPreferred = true;
-                        JLOG(this->j_.trace()) << "Counting for " << currentLedger
-                                         << " not " << v.ledgerID();
-                    }
-
-                    if (countPreferred)
-                        ret[currentLedger]++;
-                    else
-                        ret[v.ledgerID()]++;
-                }
+        // No trusted validations to determine branch
+        if (preferredSeq == Seq{0})
+            return std::make_pair(preferredSeq, preferredID);
+
+        Seq currSeq = currLedger.seq();
+        ID currID = currLedger.id();
+
+        // If we are the parent of the preferred ledger, stick with our
+        // current ledger since we might be about to generate it
+        if (preferredSeq == currSeq + Seq{1})
+        {
+            for (auto const& it : lastLedger_)
+            {
+                Ledger const& ledger = it.second;
+                if (ledger.seq() == preferredSeq &&
+                    ledger.id() == preferredID && ledger[currSeq] == currID)
+                    return std::make_pair(currSeq, currID);
+            }
+        }
+
+        // A ledger ahead of us is preferred regardless of whether it is
+        // a descendent of our working ledger or it is on a different chain
+        if (preferredSeq > currSeq)
+            return std::make_pair(preferredSeq, preferredID);
+
+        // Only switch to earlier sequence numbers if it is a different
+        // chain to avoid jumping backward unnecessarily
+        if (currLedger[preferredSeq] != preferredID)
+            return std::make_pair(preferredSeq, preferredID);
+
+        // Stick with current ledger
+        return std::make_pair(currSeq, currID);
+    }
+
+    /** Get the ID of the preferred working ledger that exceeds a minimum valid
+        ledger sequence number
+
+        @param currLedger Current working ledger
+        @param minValidSeq Minimum allowed sequence number
+
+        @return ID Of the preferred ledger, or currLedger if the preferred ledger
+                   is not valid
+    */
+    ID
+    getPreferred(Ledger const& currLedger, Seq minValidSeq)
+    {
+        std::pair<Seq, ID> preferred = getPreferred(currLedger);
+        if(preferred.first >= minValidSeq && preferred.second != ID{})
```

Comment:
@nbougalis, if the plan is too deprecate the `base_uint` default constructor behavior, I can rewrite this to use `boost::optional` instead. I was tempted by this approach since a ledger hash of 0 already has a special meaning.

Reply:
I'm fine with using zero (as it does have a special meaning, as you point out). I just want to make sure that if we remove the "zero init on default construction" this code won't accidentally break.

Reply:
I'll make it explicit as `ID{0}`.